File: scala/binary-search/src/main/scala/BinarySearch.scala
import scala.annotation.tailrec

object BinarySearch {
  @tailrec
  def aux(input: List[Int], lo: Int, hi: Int, target: Int): Option[Int] = {
    val mid: Int = (lo + hi) / 2
    if (lo > hi) None
    else if (input(mid) < target) aux(input, mid + 1, hi, target)
    else if (input(mid) > target) aux(input, lo, mid - 1, target)
    else Some(mid)
  }

  def find(input: List[Int], target: Int) = aux(input, 0, input.length - 1, target)
}

File: csharp/binary-search/BinarySearch.cs
public static class BinarySearch
{
    public static int Find(int[] input, int target)
    {
        int lo = 0;
        int hi = input.Length - 1;
        while (lo <= hi)
        {
            int mid = (lo + hi) / 2;
            if (target > input[mid])
                lo = mid + 1;
            else if (target < input[mid])
                hi = mid - 1;
            else
                return mid;
        }

        return -1;
    }
}
